ivy/functional/frontends/numpy/func_wrapper.py

```diff
@@ -0,0 +1,95 @@
+import ivy
+import functools
+from typing import Callable
+
+
+def _is_same_kind_or_safe(t1, t2):
+    if ivy.is_float_dtype(t1):
+        return ivy.is_float_dtype(t2) or ivy.can_cast(t1, t2)
+    elif ivy.is_uint_dtype(t1):
+        return ivy.is_uint_dtype(t2) or ivy.can_cast(t1, t2)
+    elif ivy.is_int_dtype(t1):
+        return ivy.is_int_dtype(t2) or ivy.can_cast(t1, t2)
+    elif ivy.is_bool_dtype(t1):
+        return ivy.is_bool_dtype(t2) or ivy.can_cast(t1, t2)
+    raise ivy.exceptions.IvyException(
+        "dtypes of input must be float, int, uint, or bool"
+    )
+
+
+def _assert_args_and_fn(args, dtype, fn):
+    ivy.assertions.check_all_or_any_fn(
+        *args,
+        fn=fn,
+        type="all",
+        message="type of input is incompatible with dtype: {}".format(dtype),
+    )
+
+
+def handle_numpy_casting(num_args: int = 1) -> Callable:
+    def inner_decorator(fn: Callable) -> Callable:
+        @functools.wraps(fn)
+        def new_fn(*args, casting="same_kind", dtype=None, **kwargs):
+            """
+            Check numpy casting type.
+
+            Parameters
+            ----------
+            args
+                The arguments to be passed to the function.
+
+            kwargs
+                The keyword arguments to be passed to the function.
+
+            Returns
+            -------
+                The return of the function, or raise IvyException if error is thrown.
+            """
+            ivy.assertions.check_elem_in_list(
+                casting,
+                ["no", "equiv", "safe", "same_kind", "unsafe"],
+                message="casting must be one of [no, equiv, safe, same_kind, unsafe]",
+            )
+            args = list(args)
+            if casting == "no" or casting == "equiv":
+                _assert_args_and_fn(
+                    args[:num_args],
```

Comment:
this will not work for functions like `numpy.concatenate`, in which the first argument is a list of arrays. You will need a more general approach which uses the `ivy.nest` module to recursively find the indices of all arrays in the input `*args` and `**kwargs`. something like:
```
arg_arrays = ivy.multi_index_nest(ivy.nested_argwhere(args, ivy.is_ivy_array), args)
kwarg_arrays = ivy.multi_index_nest(ivy.nested_argwhere(kwargs, ivy.is_ivy_array), kwargs)
arrays = arg_arrays + kwarg_arrays
```